File: src/common/user11681/wheel/util/ObservableMap.java
package user11681.wheel.util;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.function.BiConsumer;

public class ObservableMap<K, V> extends HashMap<K, V> {
    protected final List<BiConsumer<K, V>> observers = new ArrayList<>();

    @Override
    public V put(K key, V value) {
        V old = super.put(key, value);
        this.observers.forEach(observer -> observer.accept(key, value));

        return old;
    }

    public void observe(BiConsumer<K, V> observer) {
        this.observers.add(observer);
    }

    public void all(BiConsumer<K, V> action) {
        this.observe(action);
        this.forEach(action);
    }
}

File: src/net/auoeke/wheel/util/ObservableMap.kt
package net.auoeke.wheel.util

class ObservableMap<K, V> : HashMap<K, V>() {
    private val observers: MutableList<(K, V) -> Unit> = ArrayList()

    override fun put(key: K, value: V): V? {
        val old = super.put(key, value)
        this.observers.forEach {it(key, value)}

        return old
    }

    fun observe(observer: (K, V) -> Unit) {
        this.observers.add(observer)
    }

    fun all(action: (K, V) -> Unit) {
        this.observe(action)
        this.forEach(action)
    }
}
